### DISCUSSION TITLE
Documentation Discussion: Benefits of XState

### DISCUSSION THREAD

**Discussion by with-heart** [statelyai/xstate#2182] [General]
Today has been a great day for documentation at XState. There was an incredible amount of discussion in the [#documentation channel on Discord](https://discord.gg/hNcAwXJFPK). One of the key inights in my view was that we need to start tracking what it is we'd like to document, so that we can start working on outlines and actually committing these plans and ideas.

As part of that, I wanted to open up this discussion to start tracking all of the things we view as the benefits of using XState. There are a ton of benefits and it's hard to follow a lot of what was discussed on Discord, so I was hoping we could use this discussion as a central tracking place.

**Comment by ChrisShank** [comment]
I think the inverse of this question can also be insightful here! Paraphrasing a question that David asked

> What would have helped avoid that obstacle to get [xstate] adoption?

I had a few thoughts here:

- For me I think the first obstacle is how different statecharts and actors required you to think. It requires lots of persistence to change the fundamental mental models of how you reason and design about computer systems.  There is something really tacit about this process too, someone new to these concepts really has to experience the benefits of modeling behavior before they can fully understand the benefits.
- Another observation i have noticed is that modeling behavior with a state machines can be an uncomfortable process for someone new. It’s very different to think about behavior in a declarative manner as opposed to the imperative booleans, if statements, and  for loops they are so use too!

**Reply by ChrisShank** [reply]
this was `@dhirj`'s response
> i think the hardest part about convincing my old team was not just that the learning curve was higher than expected, but that it wasn't easy to figure out strong and scalable patterns that could be used across the enterprise level project. I myself was still trying to figure out the best way to do a number of things, and couldn't find any readily available materials for many of my questions either. I think we can address this with strong complex examples, and a best practices section.

**Comment by ChrisShank** [comment]
I think statecharts provide a fundamental approach to structuring the behavior of your application. Before that behavior was distributed across your entire application; located across multiple components, state management libraries, ect. With `xstate` your behavior becomes colocated into a singular place that is easy to follow (even if it is a large statechart). In fact, @johnyanarella even connected this organizational aspect of statecharts back to their initial development:

> Interesting how that's program logic taking on exactly the same scattered form as the specification documents that inspired Harel to invent Statecharts in the first place! (per the "Statecharts in the making" paper)

Now here is the cool part, from that organization and the declarative nature of statecharts emerges visualization, better reusability/composition, easier understanding, easier testing, easier time debugging and refactoring!

**Comment by ChrisShank** [comment]
The distributed nature of the actor model is a **huge** benefit, and something i think xstate has a clear advantage over (all?) other js state management libraries! ([my original realization](https://discord.com/channels/795785288994652170/800812250306183178/839406611020644353))

With first class support for web workers, web sockets, ect. i think a lot of possibilities will be unlocked!

**Reply by ChrisShank** [reply]
The implications of a distributed system are not just about mutli-threaded environments though (even though i sort of alluded too that above). It is about coming to terms with that fact that there is no single source of truth in a browser environment and that your web application is just an actor spawned within many other stateful systems. The browser tab is stateful (e.g tab visibility, URL, requestAnimationFrame, ect.), the DOM is stateful (e.g focus, accessibility, input values, transitions)... 

The actor model allows those systems to retain control of the state they own and allow your web app to reach a state of *eventual consistently* with them. I realized in my research about client-side routing that significant complexity can arise when web apps pretend that they are in full control of state that they dont own (and i think this extends to much more than just routing, just look at Redux’s philosophy to be the single source of truth!)

**Comment by with-heart** [comment]
Declarative modeling of code/data

> Procedural thinking is sort of like trying to cook by reasoning about the molecules of the ingredients. It is almost too low-level of an abstraction to think at for what we are trying to describe. In school and work, we are rarely taught about any higher-level abstractions to reason about. (i.e. usually students only exposure to state machines is parsing binary strings :yawning_face:). So we are ingrained with reasoning at the wrong level of abstraction because we have no other alternatives (except maybe pseudo code, which is still very procedural haha)
—@ChrisShank

> I think there's a few levels of abstraction here:
> 
> 1. Declaratively defining data (schemas, etc.)
> 2. Declaratively defining data transformation (SQL, FP, etc.)
> 3. Declaratively defining how data changes over time in response to events (state machines, etc.)
>
> I think most programs fit under all 3 levels
> Except we tend to be lazy and not define it declaratively, but rather "whatever code will make it work"
—@davidkpiano

**Reply by ChrisShank** [reply]
> This reminds me of @jonathanj's "thinking about code" comment earlier, and how whiteboarding evokes a different frame of mind (and more readily activates spatial reasoning). Thinking about procedures vs thinking about a system of things and how they overlap, connect and change over time.
> Q: Is that procedural thinking innate, or just something ingrained now as a consequence of the languages used to teach programming? (I think the latter, and that's the obstacle you have to overcome.)
> In that edX course, Harel builds his examples around familiar things in the real world (ex. a baby asleep, awake, sitting, standing) when he teaches statecharts. I think the key is connecting "states" to what people might know as "modes" (park, neutral, drive, reverse) and already use in their reasoning.
> — @johnyanarella

**Comment by ChrisShank** [comment]
Another benefit of xstate that I am starting to explore is granular logging of my web apps behavior. When debugging we want to know the path someone too to get into a certain state. When trying to understand how users use our applications we want to be able to track how they are navigating through our app. If the behavior of your app is dispersed among multiple components and state management libraries trying to log what errors and user interactions is non-trivial. If our behavior can modeled using a state chart (or actor system) logging changes to the behavior our app becomes (almost??) trivial because we have explicitly defined the states and transitions we can be in.

I think there was a youtube video where @davidkpiano talks about this?

**Reply by ChrisShank** [reply]
> Reproducing customer-reported bugs can be difficult even if you have their exact current data, because the problem may depend on a specific sequence of operations, or even the timing of asynchronous responses. We need not just snapshots, but records of change over time, to understand

https://asolove.medium.com/what-are-the-important-problems-in-ui-engineering-8b7f8b305611

**Comment by ChrisShank** [comment]
In the discord we had a couple of relevant discussions:

- do statecharts make it easier to deal with changing requirements? [link](https://discord.com/channels/795785288994652170/795785288994652174/845146876642459708)
- xstate is an state orchestration library not necessarily a state management library. I learned that the differences are subtle! [source](https://discord.com/channels/795785288994652170/795785288994652174/845722604043108362)
- why do developers keep creating their own abstraction? Statecharts provide a robust abstraction than can be used almost anywhere. [source](https://discord.com/channels/795785288994652170/795785288994652174/845760580673077249)
